File: atcoder/zone2021_e_Dijkstra.cpp
// https://atcoder.jp/contests/zone2021/tasks/zone2021_e
#include<iostream>
// #include<algorithm>
// #include<string>
// #include<numeric>
// #include<vector>
// #include<map>
// #include<tuple>
// #include<set>
// #include<queue>
// #include<deque>
// #include<regex>
// #include<bitset>
// #include<iomanip>
// #include<complex>
// #include<stack>
// #include<functional>

#include<bits/stdc++.h>
using namespace std;
// #include <atcoder/all>
// using namespace atcoder;
#define rep(i,n) for (int i = 0; i < (n); ++i)
#define drep(i,n) for(int i = (n-1); i >= 0; i--)
#define all(v) (v).begin(),(v).end()
#define maxs(x,y) (x = max(x,y))
#define mins(x,y) (x = min(x,y))
template <class T> bool chmax(T &a, const T &b) { if (a < b) { a = b; return 1; } return 0; }
template <class T> bool chmin(T &a, const T &b) { if (b < a) { a = b; return 1; } return 0; }
template <class T> T gcd(T a, T b) { return b ? gcd(b, a % b) : a; }
template <class T> T lcm(T a, T b) { return a/gcd(a,b)*b; }
typedef pair<int, int> P;
typedef long long ll;
const int INF = 1001001001;
const ll LINF = 1001002003004005006ll;
const ll MOD = 1e9+7;

int main() {
    cin.tie(nullptr);
    ios::sync_with_stdio(false);
    int r, c;
    cin >> r >> c;
    vector<vector<int>> a(r, vector<int>(c-1));
    vector<vector<int>> b(r-1, vector<int>(c));
    rep(i,r) rep(j,c-1) cin >> a[i][j];
    rep(i,r-1) rep(j,c) cin >> b[i][j];

    priority_queue<P, vector<P>, greater<P>> q;
    vector<int> dist(r*c*2, INF);
    auto push = [&](int i, int j, int k, int x) {
        int v = i*c*2 + j*2 + k;
        if (dist[v] <= x) return;
        dist[v] = x;
        q.emplace(x, v);
    };
    push(0,0,0,0);
    while (!q.empty()) {
        auto [x, v] = q.top(); q.pop();
        if (dist[v] != x) continue;
        int i = v/(c*2);
        int j = v/2%c;
        int k = v%2;
        if (k==0) {
            if (j+1 < c) push(i, j+1, k, x+a[i][j]);
            if (j-1 >=0) push(i, j-1, k, x+a[i][j-1]);
            if (i+1 < r) push(i+1, j, k, x+b[i][j]);
            push(i, j, 1, x+1);
        } else {
            if (i-1 >=0) push(i-1, j, k, x+1);
            push(i, j, 0, x);
        }
    }
    int ans = dist[(r*c-1)*2];
	cout << ans << endl;
	return 0;
}

File: atcoder/zone2021_e_Dijkstra.py
# https://atcoder.jp/contests/zone2021/tasks/zone2021_e

from scipy.sparse.csgraph import shortest_path
from scipy.sparse import csr_matrix
R, C = map(int, input().split())
data = []
row = []
column = []
for i in range(R):
    a = list(map(int, input().split()))
    for j in range(C - 1):
        x = i * C + j
        y = i * C + j + 1
        row += [x, y]
        column += [y, x]
        data += [a[j], a[j]]
for i in range(R - 1):
    a = list(map(int, input().split()))
    for j in range(C):
        x = i * C + j
        y = i * C + j + C
        row += [x, y + R * C]
        column += [y, x + R * C]
        data += [a[j], 1]
for i in range(R):
    for j in range(C):
        x = i * C + j
        y = x + R * C
        row += [x, y]
        column += [y, x]
        data += [1, 0]
# print(data)
g = csr_matrix((data, (row, column)))
print(int(shortest_path(g, indices=0)[R * C - 1]))

# TLE
# import heapq
# INF = 1001001001

# R, C = map(int, input().split())
# A = [[int(i) for i in input().split()] for _ in range(R)]
# B = [[int(i) for i in input().split()] for _ in range(R-1)]
# # print(R, C)
# # print(A)
# # print(B)

# dist = [[[INF]*2 for _ in range(C)] for _ in range(R)]
# # print(dist)

# q = [(0,0,0,0)]
# while len(q)>0:
#     cost, r, c, f = heapq.heappop(q)
#     # print(cost, r, c, f)
#     if dist[r][c][f] <= cost: continue
#     dist[r][c][f] = cost
#     if r==R-1 and c==C-1 and f==0:
#         ans = cost
#         break
#     if f==0:
#         if c<C-1: heapq.heappush(q, (cost+A[r][c], r, c+1, f))
#         if c>0:   heapq.heappush(q, (cost+A[r][c-1], r, c-1, f))
#         if r<R-1: heapq.heappush(q, (cost+B[r][c], r+1, c, f))
#         heapq.heappush(q, (cost+1, r, c, 1))
#     else:
#         if r>0: heapq.heappush(q, (cost+1, r-1, c, f))
#         heapq.heappush(q, (cost, r, c, 0))

# print(ans)
